### ISSUE TITLE
Support HTTP PATCH

### ISSUE THREAD

**Issue by ksuderman** [http-builder-ng/http-builder-ng#118]
[RFC 5789](https://tools.ietf.org/html/rfc5789) adds the HTTP `PATCH` verb which is used in several places by the GitHub API to update resources.  See for example [Update a Reference](https://developer.github.com/v3/git/refs/#update-a-reference).

If I have some spare time I might be able to contribute a PR, but before starting such a task I want to make sure I understand the scope of the changes.  A quick analysis shows that at least the following would need to be done:

1. Add PATCH to `HttpVerb` enum
1. Add `doPatch` methods to:
    * JavaHttpBuilder
    * ApacheHttpBuilder
    * OkHttpBuilder
1. Add the ten `patch` methods to HttpBuilder
    * `patch(): Object`
    * `patch(Closure): Object`
    * `patch(Consumer<HttpConfig>): Object`
    * `patch(Class<T>,Closure): T` 
    * `patch(Class<T>, Consumer<HttpConfig>) : T` 
    * `patchAsync() : CompletableFuture<Object>` 
    * `patchAsync(Closure) : CompletableFuture<Object>` 
    * `patchAsync(Consumer<HttpConfig>) : CompletableFuture<Object>` 
    * `patchAsync(Class<T>, Closure) : CompletableFuture<T>` 
    * `patchAsync(Class<T>, Consumer<HttpConfig>) : CompletableFuture<T>` 
1. Spock unit tests for all of the above.
1. Updated documentation for all of the above.

Am I missing anything else?

Edit: there are 10 methods that need to be added to `HttpBuilder`.

**Comment by cjstehno** [comment]
Sounds like that would do it. We considered PATCH support but didn't think anything used it. Please base your branch off of the `development` branch.

Luckily, I did support PATCH in the Ersatz server library for testing. ☺️

Thanks and if it becomes too much effort, let us know and we can take it over. Sounds like it would be useful.

**Comment by ksuderman** [comment]
Actually, it turned out to be relatively easy.  I simply copied everything related to the PUT method and did a few find/replace operations

1. `put(` --> `patch(`
1. `putAsync(` --> `patchAsync(`
1. `PUT` --> `PATCH`

That seemed to take care of 90% of the work.  Of course, since this is a big copy/change/paste there are plenty of opportunities for errors and repeating typos.

I have created a PR for comments and discussion, but you probably don't want to accept it as is:

* I bumped the version in `build.gradle` to 0.16.0 so I could install and test it locally without stomping on the 0.15.0 release I also have installed.
* I wasn't sure what to do with the copyright notice at the top of the files I added.  I have IntelliJ configured to automatically insert an Apache license header with my name in the copyright whenever I create a new file.  Similarly it also adds my name as the @author of classes I create.  I can go through as massage these as needed.

Other minor tweaks.

* .gitignore had `.idea` and `*.iml` listed twice but was missing `*.ipr` (IDEA project) and `*.iws` (IDEA workspace)

**Comment by cjstehno** [comment]
I should be able to review the pull request tomorrow.

**Comment by ksuderman** [comment]
I knew it seemed too easy.  

There is one major caveat that will definitely need to be documented somewhere: the [HttpURLConnection.setRequestMethod(String)](https://docs.oracle.com/javase/8/docs/api/java/net/HttpURLConnection.html#setRequestMethod-java.lang.String-) method throws a `ProtocolException` if the method is not one of the standard HTTP/1.1 verbs and PATCH is not one of the standard HTTP/1.1 verbs.

This is only a problem for the JavaHttpBuilder, both the ApacheHttpBuilder and OkHttpBuilder accept non-standard request methods and work as desired.

For the JavaHttpBuilder case there are two options:

1. fail fast and throw an exception in the `JavaHttpBuilder.doPatch` method since we know that the `setRequestMethod` is going to throw an exception.
1. there is a hack-around to set a `X-HTTP-Method-Override` header and then send the request as a `POST`. We would need to add some logic around line 58 of the JavaHttpBuilder.java file for this.

I've added some comments and a //TODO marker to the PR.

**Comment by cjstehno** [comment]
Yeah there are some hidden "gems" in these libraries we've had to work around. I'm not really liking option 2 since it's not really an expected behavior. I'd say option 1 and just throw an UnsupportedOperationException - it can be documented that PATCH is not supported by the core client. Not ideal but...

**Comment by ksuderman** [comment]
I'll make the change and throw an UnsupportedOperationException.

It might be worth considering making `ApacheHttpBuilder` the core client rather than `JavaHttpBuilder`, but that discussion probably belongs in its own issue.

**Comment by cjstehno** [comment]
The "core" library is actually poorly named... it should really be called "java-standard" or something but that's too long. It's meant to be the core of the library itself, but also a dependency-less implementation of the library (i.e. standard java).

Another note: Since the core library does not support PATCH you will want to cut the core PATCH tests off from the test kit and just test each `patch` method to ensure that they throw the expected exception.

**Comment by cjstehno** [comment]
I reviewed the current PR and it looks good (see PR comments). 

Also, I just pushed copyright header and doc updates to the `development` branch - please merge them into your branch before moving forward. Thanks.

**Comment by cjstehno** [comment]
The pull request has been merged into the `development` branch and the 0.16 release will be cut within the next few days.

**Comment by cjstehno** [comment]
This is in the 0.16.0 release, which is now out in the wild.

**Comment by ksuderman** [comment]
Excellent!  Thank you.

Of course, as I was pulling in the new version I noticed a small typo on line 101 of JavaHttpPatchSpec:

```groovy
http.patchAsync(config)exceptionally { e ->
```

should be

```groovy
 http.patchAsync(config).exceptionally { e ->
```

The code works as desired due to Groovy's method call chaining mechanism, but the missing dot between `(config)` and `exceptionally` is like fingernails on a blackboard for my OCD now that I've seen it.  The issue (if you want to call it that) likely doesn't warrant a pull-request, but someone might want to add the missing dot before the next release.

**Comment by cjstehno** [comment]
Ha... I will update it in the development branch.

**Comment by danielpsf** [comment]
Is PATCH still not supported by the core library?

I'd love to use http-builder-ng instead of rest-template (Spring) or http-builder (Apache).

**Comment by cjstehno** [comment]
As of Java 8 this is still true... not sure about Java 9, but we don't support that yet anyway. 

We support three underlying client implementations. Why not just use the OkHttp or Apache implementation rather than the core? They both support PATCH.

**Comment by danielpsf** [comment]
You mean, instead of using core integrate with OkHTTP, like this [doc](https://http-builder-ng.github.io/http-builder-ng/asciidoc/html5/#_client_library_integration) suggest.

Well. It make sense to me. Will try...

Would it be a good idea to mention this workaround on this part of [the doc](https://http-builder-ng.github.io/http-builder-ng/asciidoc/html5/#_unsupported_features) as well?

**Comment by cjstehno** [comment]
Sort of... the Apache and OkHttp clients are already supported implementations (e.g. https://http-builder-ng.github.io/http-builder-ng/docs/javadoc/groovyx/net/http/OkHttpBuilder.html) - see also https://http-builder-ng.github.io/http-builder-ng/asciidoc/html5/#_add_library_to_project for how to select the underlying client.
